File: src/openmp/f90/hello_world_openmp.f
C******************************************************************************
C FILE: hello_world_openmp.f
C AUTHOR: Blaise Barney  5/99
C LAST REVISED: Neal Davis  2/14
C******************************************************************************

      PROGRAM HELLO
      
      INTEGER NTHREADS, TID, OMP_GET_NUM_THREADS,
     +        OMP_GET_THREAD_NUM
      
      PRINT *, 'Fortran90 OpenMP minimal working example'
      
C     Fork a team of threads giving them their own copies of variables
!$OMP PARALLEL PRIVATE(NTHREADS, TID)
C     Obtain thread number
      TID = OMP_GET_THREAD_NUM()
      
C     Output number of threads
      IF (TID .EQ. 0) THEN
            NTHREADS = OMP_GET_NUM_THREADS()
            PRINT *, 'Number of available processors =', NTHREADS
      END IF
      
      PRINT *, 'Process number ', TID, ' branching off.'
      
C     All threads join master thread and disband
!$OMP END PARALLEL
      
      END

File: src/openmp/c/hello_world_openmp.c
#include <stdio.h>
#include <omp.h>

int main(int argc, char *argv[]) {
    int     rank_id;
    double  start_time, wall_time;
    
    printf("C OpenMP minimal working example\n");
    int num_ranks = omp_get_num_procs();
    printf("Number of available processors = %d.\n", num_ranks);
    int num_threads = omp_get_max_threads();
    printf("Number of available threads = %d.\n", num_threads);
    
    start_time = omp_get_wtime();
#pragma omp parallel private (rank_id)
    {
        rank_id = omp_get_thread_num();
        printf("\tProcess number %d branching off.\n", rank_id);
    }
    
    wall_time = omp_get_wtime() - start_time;
    
    printf("Elapsed wallclock time = %8.6fs.\n", wall_time);
    return 0;
}

